### ISSUE TITLE
Suggested modification to use format conversion field

### ISSUE THREAD

**Issue by mwtoews** [ikamensh/flynt#106]
The [Format String Syntax](https://docs.python.org/3/library/string.html#format-string-syntax) includes a conversion character, which can be one of:

- `s` to call `str()`
- `r` to call `repr()`
- `a` to call `ascii()`

E.g. take a look at the following:
```python
import numpy as np  # ensure this is version >=1.14

x = np.float32(101.1)
# bad output: 101.0999984741211
print("{}".format(x))
# good output: 101.1
print("{!s}".format(x))
print("{}".format(str(x)))
print("%s" % x)
print("%s" % str(x))
print("{!r}".format(x))
print("{}".format(repr(x)))
print("%r" % x)
print("%s" % repr(x))
```
(side note about the "bad output" is that is due to conversion from float32 to Python's native 64-bit representation, and we never want to see this as it's a bad conversion).

Currently, flynt v.0.65 doesn't attempt to modify the conversion character of the format string. For example:

- `"{}".format(str(x))` (and `"%s" % str(x)`) should be translated to `f"{x!s}"` rather than `f"{str(x)}"`
- `"{}".format(repr(x))` (and `"%s" % repr(x)`) should be translated to `f"{x!r}"` rather than `f"{repr(x)}"`.
- `"{}".format(ascii(x))` (and `"%s" % ascii(x)`) should be translated to `f"{x!a}"` rather than `f"{ascii(x)}"`.

The motivation behind this suggestion is that the resulting conversion is shorter, preserves output intent (i.e. no unexpected conversions, demonstrated with numpy), and tests with `%timeit` consistently show measurable efficiencies with format strings that directly use the conversion character. This is shown with:
```python
In [1]: from dis import dis

In [2]: dis(compile('f"{x!s}"', '', 'exec'))
  1           0 LOAD_NAME                0 (x)
              2 FORMAT_VALUE             1 (str)
              4 POP_TOP
              6 LOAD_CONST               0 (None)
              8 RETURN_VALUE

In [3]: dis(compile('f"{str(x)}"', '', 'exec'))
  1           0 LOAD_NAME                0 (str)
              2 LOAD_NAME                1 (x)
              4 CALL_FUNCTION            1
              6 FORMAT_VALUE             0
              8 POP_TOP
             10 LOAD_CONST               0 (None)
             12 RETURN_VALUE
```
As you can see, more less processing is required when format conversions are embedded.

**Comment by ikamensh** [comment]
Hi, thanks for this suggestion.

TLDR: I'm open to having this in flynt, but unlikely to find time to implement it as of now. Unless it's provably safe, this should be under --aggressive flag or some new flag and not a default behavior.

The main concern of `flynt` is to convert .format calls and % formatting to f-strings. As I understand, any code that has .format() call could together with e.g. ascii() was not using the possible formatting notation before. 

So this is like an opportunity to solve one more problem, namely not using the !r, !s, !a formatting modifiers. I suspect that many programmers actually don't know about them at all.

## Implementation
* Implementation-wise, this could be a part of NodeTransformer visit to all AST Call nodes.
* Putting it under some flag or not: this can be default behavior of `flynt` iff under the hood exactly the same mechanisms are guaranteed to be called.
* !s formatting specifier - is it any different from not having `!s`? I thought that everything without a specifier will be treated as !s; if it is so, maybe it should always be omitted?

**Comment by mwtoews** [comment]
I'd imagine transferring simple `str(x)` their equivalent `!s` conversion field is safe, as this is the expected behaviour since [Python 3.0](https://docs.python.org/3.0/library/string.html#format-string-syntax) (and possibly [Python 2.6](https://docs.python.org/2.6/library/string.html#formatspec), but this only had `!r` and `!s`). But yes, `f{x!s}` is different than `f{x}`.

The numpy demo above is not obvious, so here is a clearer example of the class behaviour:
```python
class Foo:
    def __format__(self, format_spec):
        return f"format with spec: {format_spec!r}"
    def __str__(self):
         return "str"
    def __repr__(self):
         return "repr–"

obj = Foo()
print(f"{obj}")
print(f"{obj:some things}")
print(f"{obj!s}")
print(f"{obj!r}")
print(f"{obj!a}")
```
prints
```
format with spec: ''
format with spec: 'some things'
str
repr–
repr\u2013
```
I'll admit that I thought `!s` was the same as without too, but internally it calls either `__str__` or `__format__` methods, where implemented. And the content of `format_spec` (after `:`) is a whole other thing that has nothing to do with this feature.

**Comment by ikamensh** [comment]
yeah as far as I can tell it would be safe to use these methods. Here is an attempt at a proof by chatgpt codex:
```
"""Demonstrations for issue #106 about conversion field equivalence.

The f-string conversion specifiers ``!s``, ``!r`` and ``!a`` internally call
``str()``, ``repr()`` and ``ascii()`` respectively.  This script shows:

* objects with and without ``__format__`` behave identically when using ``!s``,
  ``!r`` and ``!a`` compared with explicit function calls;
* overriding the built-in names ``str``/``repr``/``ascii`` breaks this
  equivalence.

Python's bytecode for ``f"{x!s}"`` uses the ``FORMAT_VALUE`` opcode with a
conversion flag. This always resolves to the real built-in functions even if the
names ``str`` or ``repr`` are shadowed in the local scope. Therefore, barring
such shadowing, ``f"{x!s}"`` is equivalent to ``f"{str(x)}"`` for all objects.

Run this file directly to print the examples.
"""


class Custom:
    """Defines ``__format__`` in addition to ``__str__`` and ``__repr__``."""

    def __format__(self, spec: str) -> str:
        return f"formatted({spec})"

    def __str__(self) -> str:
        return "string"

    def __repr__(self) -> str:
        return "repr"


class WithoutFormat:
    """Lacks a ``__format__`` implementation."""

    def __str__(self) -> str:
        return "plain-str"

    def __repr__(self) -> str:
        return "plain-repr"


def demonstrate(obj) -> None:
    print("object:", obj)
    print("default f-string:", f"{obj}")
    print("f'{obj!s}':", f"{obj!s}")
    print("f'{str(obj)}':", f"{str(obj)}")
    print("f'{obj!r}':", f"{obj!r}")
    print("f'{repr(obj)}':", f"{repr(obj)}")
    print("f'{obj!a}':", f"{obj!a}")
    print("f'{ascii(obj)}':", f"{ascii(obj)}")
    print()


def shadowed_demo() -> None:
    print("Shadowing built-ins breaks equivalence:")

    def str(x):  # noqa: A001
        return "shadowed-str"

    def repr(x):  # noqa: A001
        return "shadowed-repr"

    def ascii(x):  # noqa: A001
        return "shadowed-ascii"

    obj = Custom()
    print("f'{obj!s}':", f"{obj!s}")
    print("f'{str(obj)}':", f"{str(obj)}")
    print("f'{obj!r}':", f"{obj!r}")
    print("f'{repr(obj)}':", f"{repr(obj)}")
    print("f'{obj!a}':", f"{obj!a}")
    print("f'{ascii(obj)}':", f"{ascii(obj)}")
    print()


def main() -> None:
    values = [123, 45.6, "text", Custom(), WithoutFormat()]
    for value in values:
        demonstrate(value)
    shadowed_demo()


if __name__ == "__main__":
    main()
```
